stone/target/js_types.py

```diff
@@ -0,0 +1,213 @@
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import argparse
+import json
+import six
+import sys
+import os
+import re
+
+from stone.data_type import (
+    is_user_defined_type,
+    is_union_type,
+    is_struct_type,
+    is_void_type,
+    unwrap,
+)
+from stone.generator import CodeGenerator
+from stone.target.js_helpers import (
+    fmt_type,
+)
+
+
+_cmdline_parser = argparse.ArgumentParser(prog='js-types-generator')
+_cmdline_parser.add_argument(
+    'filename',
+    help=('The name to give the single Javascript file that is created and '
+          'contains all of the JSDoc types.'),
+)
+_cmdline_parser.add_argument(
+    '-e',
+    '--extra-arg',
+    action='append',
+    type=str,
+    default=[],
+    help=("Additional properties to add to a route's argument type based "
+          "on if the route has a certain attribute set. Format (JSON): "
+          '{"match": ["ROUTE_ATTR", ROUTE_VALUE_TO_MATCH], '
+          '"arg_name": "ARG_NAME", "arg_type": "ARG_TYPE", '
+          '"arg_docstring": "ARG_DOCSTRING"}'),
+)
+
+_header = """\
+// Auto-generated by Stone, do not modify.
+/**
+ * An Error object returned from a route.
+ * @typedef {Object} Error
+ * @property {string} error_summary - Text summary of the error.
+ * @property {T} error - The error object.
+ * @property {UserMessage} user_message - User-friendly error message.
+ * @template T
```

Comment:
JSDoc itself doesn't seem to recognize `@template`, but it's a JSDoc tag used by Google Closure Compiler to denote generics and keeping it around doesn't hurt doc generation. Googling suggests that some IDEs, like WebStorm, recognize this tag.